browser/src/main/java/org/netbeans/html/presenters/browser/Browser.java

```diff
@@ -431,27 +472,73 @@ private void emitScript(Writer w, String prefix, String id) throws IOException {
                     + "    request.onreadystatechange = function() {\n"
                     + "      if (this.readyState!==4) return;\n"
                     + "      try {\n"
+            );
+            if (Browser.this.config.debug) {
+                w.write(""
+                    + "        console.log('...GET[' + counter + '] got something ' + this.responseText.substring(0,80));\n"
                     + "        var cmd = document.getElementById('cmd');\n"
                     + "        if (cmd) cmd.innerHTML = this.responseText.substring(0,80);\n"
+                );
+            }
+            w.write(""
                     + "        (0 || eval)(this.responseText);\n"
                     + "      } catch (e) {\n"
                     + "        console.warn(e); \n"
                     + "      } finally {\n"
-                    + "        waitForCommand();\n"
+                    + "        waitForCommand(counter + 1);\n"
                     + "      }\n"
                     + "    };\n"
                     + "    request.send();\n"
                     + "  } catch (e) {\n"
                     + "    console.warn(e);\n"
-                    + "    waitForCommand();\n"
+                    + "    waitForCommand(counter + 1);\n"
                     + "  }\n"
                     + "}\n"
-                    + "waitForCommand();\n"
+                    + "waitForCommand(1);\n"
             );
             w.write("  </script>\n");
         }
     }
 
+    String createCallbackFn(String prefix, String id) {
+        StringBuilder sb = new StringBuilder();
+        sb.append("this.toBrwsrSrvr = function(name, a1, a2, a3, a4) {\n"
+            + "var url = '").append(prefix).append("command.js?id=").append(id).append("&name=' + name;\n"
+            + "var body = 'p0=' + encodeURIComponent(a1);\n"
+            + "body += '&p1=' + encodeURIComponent(a2);\n"
+            + "body += '&p2=' + encodeURIComponent(a3);\n"
+            + "body += '&p3=' + encodeURIComponent(a4);\n"
+            + "var request = new XMLHttpRequest();\n"
+        );
+        if (Browser.this.config.debug) {
+            sb.append(""
+            + "console.log('PUT ... ' + body.substring(0, 80));\n"
+            + "var now = new Date().getTime();\n"
+            );
+        }
+        sb.append(""
+            + "request.open('PUT', url, false);\n"
+            + "request.setRequestHeader('Content-Type', 'application/x-www-form-urlencoded; charset=utf-8');\n"
```

Comment:
Whenever JavaScript needs to invoke something in the JVM, it initiates new blocking `PUT` request to the JVM server. JVM is supposed to invoke the desired method and returns the result back to JavaScript.

The trouble starts if the JVM needs to execute some JavaScript while processing the `PUT` request. Then it returns the JavaScript snippet from this method, let's the JavaScript side process it and invoke another `PUT` request to continue. Most of this logic is handled by the `Generic` class as far as I can tell.